parlai/core/distributed_utils.py

```diff
@@ -0,0 +1,201 @@
+            "because all workers will have exactly the same minibatches, "
+            "defeating the purpose."
+        )
+
+
+def is_distributed():
+    """
+    Returns True if we are in distributed mode.
+    """
+    return TORCH_AVAILABLE and dist.is_initialized()
+
+
+def num_workers():
+    """
+    Get the total number of workers.
+    """
+    if not is_distributed():
+        return 1
+    else:
+        return dist.get_world_size()
+
+
+def is_primary_worker():
+    """
+    Returns False if we are a secondary worker. Returns True if we are either
+    (1) not in distributed mode (2) or are the primary (rank 0) worker.
+    """
+    return not is_distributed() or dist.get_rank() == 0
+
+
+def override_print(suppress=False, prefix=None):
+    """
+    Overrides the builtin print, to either mute or annotate the output of
+    individual workers. Mutes if suppress is True, and this is not the
+    primary or only worker.
+    """
+    builtin_print = builtins.print
+
+    def new_print(*args, **kwargs):
+        if suppress:
+            # do nothing
+            return
+        elif prefix:
+            return builtin_print(prefix, *args, **kwargs)
+        else:
+            # default to normal print
+            return builtin_print(*args, **kwargs)
+
+    builtins.print = new_print
+
+
+def all_gather_list(data, max_size=16384):
+    """Gathers arbitrary data from all nodes into a list.
+    Similar to :func:`~torch.distributed.all_gather` but for arbitrary Python
+    data. Note that *data* must be picklable.
+    Args:
+        data (Any): data from the local worker to be gathered on other workers
+        max_size (int, optional): maximum size of the data to be gathered
+            across workers
+    """
+
+    if not is_distributed():
+        # fall back to just keeping things basic if we're not distributed
+        return [data]
+
+    # stolen shamelessly from fairseq
+    # https://github.com/pytorch/fairseq/blob/c37250ab1c845919af721cd3f5c4cec2993aefe1/fairseq/distributed_utils.py#L116-L170
+    rank = dist.get_rank()
+    world_size = dist.get_world_size()
+
+    buffer_size = max_size * world_size
+    if (
+        not hasattr(all_gather_list, '_buffer') or
+        all_gather_list._buffer.numel() < buffer_size
+    ):
+        all_gather_list._buffer = torch.cuda.ByteTensor(buffer_size)
+
+    buffer = all_gather_list._buffer
+    buffer.zero_()
+
+    enc = pickle.dumps(data)
+    enc_size = len(enc)
+    if enc_size + 2 > max_size:
+        raise ValueError('encoded data exceeds max_size: {}'.format(enc_size + 2))
+    assert max_size < 255 * 256
+
+    buffer_rank = buffer[rank * max_size: (rank + 1) * max_size]
+    buffer_rank[0] = enc_size // 255  # this encoding works for max_size < 65k
+    buffer_rank[1] = enc_size % 255
+    buffer_rank[2: enc_size + 2] = torch.ByteTensor(list(enc))
+
+    dist.all_reduce(buffer)
+
+    result = []
+    for i in range(world_size):
```

Comment:
I notice in the fairseq implementation, they catch the unpickling error and raise their own exception. Did you intentionally omit that?

Reply:
That's newer than the version of the code I was looking at. It's a good idea; I'll add it.